docs/architecture/adr-038-state-listening.md

```diff
@@ -182,349 +139,606 @@ func (rs *Store) GetKVStore(key types.StoreKey) types.KVStore {
+    })
+}
+```
+
+We will introduce a plugin loading system that will return `(interface{}, error)`.
+This provides the advantage of using versioned plugins where the plugin interface and gRPC protocol change over time.
+In addition, it allows for building independent plugin that can expose different parts of the system over gRPC.
 
 ```go
-func NewSimApp(
-	logger log.Logger,
-	db dbm.DB,
-	traceStore io.Writer,
-	loadLatest bool,
-	appOpts servertypes.AppOptions,
-	baseAppOptions ...func(*baseapp.BaseApp),
-) *SimApp {
+func NewStreamingPlugin(name string, logLevel string) (interface{}, error) {
+    logger := hclog.New(&hclog.LoggerOptions{
+       Output: hclog.DefaultOutput,
+       Level:  toHclogLevel(logLevel),
+       Name:   fmt.Sprintf("plugin.%s", name),
+    })
+
+    // We're a host. Start by launching the streaming process.
+    env := os.Getenv(GetPluginEnvKey(name))
+    client := plugin.NewClient(&plugin.ClientConfig{
+       HandshakeConfig: HandshakeMap[name],
+       Plugins:         PluginMap,
+       Cmd:             exec.Command("sh", "-c", env),
+       Logger:          logger,
+       AllowedProtocols: []plugin.Protocol{
+           plugin.ProtocolNetRPC, plugin.ProtocolGRPC},
+    })
+
+    // Connect via RPC
+    rpcClient, err := client.Client()
+    if err != nil {
+       return nil, err
+    }
 
-	...
-
-	keys := sdk.NewKVStoreKeys(
-	authtypes.StoreKey, banktypes.StoreKey, stakingtypes.StoreKey,
-	minttypes.StoreKey, distrtypes.StoreKey, slashingtypes.StoreKey,
-	govtypes.StoreKey, paramstypes.StoreKey, ibchost.StoreKey, upgradetypes.StoreKey,
-	evidencetypes.StoreKey, ibctransfertypes.StoreKey, capabilitytypes.StoreKey,
-	)
-
-	pluginsOnKey := fmt.Sprintf("%s.%s", plugin.PLUGINS_TOML_KEY, plugin.PLUGINS_ON_TOML_KEY)
-	if cast.ToBool(appOpts.Get(pluginsOnKey)) {
-		// this loads the preloaded and any plugins found in `plugins.dir`
-		pluginLoader, err := loader.NewPluginLoader(appOpts, logger)
-		if err != nil {
-			// handle error
-		}
-
-		// initialize the loaded plugins
-		if err := pluginLoader.Initialize(); err != nil {
-			// handle error
-		}
-
-		// register the plugin(s) with the BaseApp
-		if err := pluginLoader.Inject(bApp, appCodec, keys); err != nil {
-			// handle error
-		}
-
-		// start the plugin services, optionally use wg to synchronize shutdown using io.Closer
-		wg := new(sync.WaitGroup)
-		if err := pluginLoader.Start(wg); err != nil {
-			// handler error
-		}
-	}
+    // Request streaming plugin
+    return rpcClient.Dispense(name)
+}
+
+```
 
-	...
+We will expose a `RegisterStreamingPlugin` function for the App to register streaming plugins with the App's BaseApp.
 
-	return app
+```go
+// baseapp/streaming.go
+
+// RegisterStreamingPlugin registers the ABCI streaming service provided by the streaming plugin.
+func RegisterStreamingPlugin(
+    bApp *BaseApp,
+    appOpts servertypes.AppOptions,
+    keys map[string]*types.KVStoreKey,
+    streamingService interface{},
```

Comment:
Just reading the golang one has to wonder about the choice for `streamingService any` and type casting here.  Could you provide a brief explanation of the choice in the ADR and perhaps an example usage where `any` is needed?

Reply:
I updated the language on `RegisterStreamingPlugin`. Also see https://github.com/cosmos/cosmos-sdk/pull/13473#discussion_r990118166